plugins/webservercodes/webserver_codes.go

```diff
@@ -0,0 +1,255 @@
+# List of virtualhosts for http codes collecting (each section for one virtualhost, none for disable collecting codes)
+[[webservercodes.vhosts]]
+# 'host' field should match hostname in appropriate status url
+host = "defaulthost"
+
+# telegraf user must have read permissions to this log file (shell command 'sudo adduser telegraf adm' for apache on Ubuntu)
+access_log = "/var/log/apache2/access.log"
+
+# Regular expression for fetching codes from log file strings. You can adjust regex for your custom log format
+# example for apache "common" and "combined" log formats, nginx default log format ("combined"):
+#   '\[(?P<time>[^\]]+)\] ".*?" (?P<code>\d{3})'
+regex_parsestring = '\[(?P<time>[^\]]+)\] ".*?" (?P<code>\d{3})'
+
+# plugin will count http codes in this interval. Interval must be in time.Duration format (see https://golang.org/pkg/time/#ParseDuration)
+interval = "10s"
+`
+
+func (n *Webservercodes) SampleConfig() string {
+	return sampleConfig
+}
+
+func (n *Webservercodes) Description() string {
+	return "Read webserver access log files and collects http return codes stats"
+}
+
+func (n *Webservercodes) Gather(acc plugins.Accumulator) error {
+	var wg sync.WaitGroup
+	
+    hostStats := map[string]HttpStats{}
+    errChan := make(chan error)
+    successChan := make(chan bool)
+    remainingItems := len(n.Vhosts)
+    
+    for _, vhost := range n.Vhosts {
+        if duration, err := time.ParseDuration(vhost.Interval); err == nil {
+            wg.Add(1)
+            go func(host string, logfile string, regex string, duration time.Duration, successChan chan bool, errChan chan error) {
+                defer wg.Done()
+                if stats, err := n.ParseHttpCodes(logfile, regex, duration); err == nil {
+                    hostStats[host] = *stats
+                    successChan <- true
+                } else {
+                    errChan <- err
+                }
+            }(vhost.Host, vhost.AccessLog, vhost.RegexParsestring, duration, successChan, errChan)
+        } else {
+            return err
+        }
+    }
+    for {
+        select {
+            case _ = <-successChan:
+                remainingItems--
+            case err := <-errChan:
+                return err
+        }
+        if remainingItems == 0 {
+            break;
+        }
+    }
+    
+    wg.Wait()
+	
+    for vhost, stats := range hostStats {
+        n.gatherCodes(vhost, stats, acc)
+    }
+	
+	return nil
+}
+
+func SearchStringInSlice(a string, list []string) bool {
+    for _, b := range list {
+        if b == a {
+            return true
+        }
+    }
+    return false
+}
+
+func (n Webservercodes) ParseRegex(regex string, f io.ReadSeeker) (*regexp.Regexp, error) {
+    
+    if rx, err := regexp.Compile(regex); err == nil {
+        
+        keys := rx.SubexpNames();
+        if SearchStringInSlice("time", keys) && SearchStringInSlice("code", keys) {
+            
+            reader := reverse.NewScanner(f)
```

Comment:
Because often log files can have size of several GBs, and we need to scan only last n seconds. There often are configurations without logrotating, especially for custom virtualhosts, and one file can have days or monthes of log entries inside. Also, because of unpredictable load (queries per second) where is no easy way to find out the position to read file from (if you want to use forward scan). So I decided to just scan log file backwards and to compare each log entry time.